### DISCUSSION TITLE
How can I use micropython on the Seeedstudio XIAO-RP2040?

### DISCUSSION THREAD

**Discussion by TheChymera** [micropython/micropython#11300] [General]
I got [one of these devices](https://wiki.seeedstudio.com/XIAO-RP2040-with-MicroPython) hoping to replace pyboards which are increasingly hard to get.

Their wiki, linked above, seems to be pretty fixated on a specific IDE. But since I can just mount the drive of the device I assume I should be able to copy micropython files over to it and have it run them.

I tried a really simple set-up for controlling the LEDs (`boot.py` and `main.py`), but nothing happens.
Further, after I unmount the drive whatever files I added to it just disappear.
Any ideas how I might just transfer micropython scripts for it to run?

**Comment by Josverl** [comment]
Mpremote should just work,  as a range of other tools.

https://pypi.org/project/mpremote/

**Comment by fdufnews** [comment]
> Their wiki, linked above, seems to be pretty fixated on a specific IDE.

They suggest Thonny as it is a simple IDE tailored for the newbie. You have all the tools at hand editor, and file management.

I am using for years [mpfshell](https://github.com/wendlers/mpfshell) with ESP8266, ESP32 and Raspberry Pi Pico. It would probably works with a XIAO as it is using the REPL commands.

**Comment by TheChymera** [comment]
@Josverl @fdufnews thank you for the suggestions. They indeed look preferable to thonny. But why would I need an extra application at all to just copy files to the device and let it run them when I restart it in non-drive mode? My system already sees a drive which it can mount. Is that not the drive which scripts go into?

**Reply by robert-hh** [reply]
What is the drive name that you see. It if is RPI-RP2, it's the virtual drive of the bootloader which allows to install the firmware. You have to upload MicroPython first, and then it will run.

**Reply by TheChymera** [reply]
@robert-hh indeed it is `RPI-RP2`. So I tried installing the firmware as indicated [here](http://micropython.org/download/seeed_xiao_nrf52/):

```
[dark]  ❱ ls
RPI-RP2
[dark]  ❱ cp ~/Downloads/seeed_xiao_nrf52-20230414-unstable-v1.19.1-1016-gb525f1c9e.uf2 RPI-RP2/
[dark]  ❱ ls RPI-RP2/
INDEX.HTM  INFO_UF2.TXT  seeed_xiao_nrf52-20230414-unstable-v1.19.1-1016-gb525f1c9e.uf2
[dark]  ❱ umount RPI-RP2/
[dark]  ❱ ls
<un-plug>
<re-plug>
<un-plug>
<re-plug while holding down the boot button (i.e.) to expose the storage>
[dark]  ❱ ls RPI-RP2/
INDEX.HTM  INFO_UF2.TXT
[dark]  ❱ cat RPI-RP2/INFO_UF2.TXT
UF2 Bootloader v3.0
Model: Raspberry Pi RP2
Board-ID: RPI-RP2
```

And the file is gone again. Did micropython already get installed? If so, how can I check?

**Reply by robert-hh** [reply]
You copied the wrong file. The one you copied is for a nrf52840 board, not a RP2040 board. Try the generic RPi-Pico firmware. It should fit. https://micropython.org/resources/firmware/rp2-pico-20230414-unstable-v1.19.1-1016-gb525f1c9e.uf2
When the firmware download succeed, the RP2-RPI drive disappears, and using a serial terminal or the REPL window of Thonny you should get a Python REPL prompt.
For normal operation you must not hold down the boot button when powering the board. That's "just" the sequence to get into the firmware update mode.

**Comment by fdufnews** [comment]
Contrary to CircuitPython, MicroPython does not appear as a mass storage when plugged-in in a computer.
When you plug the Pico in while holding the boot push button the drive you see is a virtual one exposed by the bootloader as explained by @robert-hh. The sole purpose of this drive is to install a firmware not any other type of file. Hence the use of another tool like Thonny, mpremote, mpfshell, ...

**Reply by TheChymera** [reply]
Oh, interesting, so there is indeed no way to “just” copy the instructions then? That's a bit sad. On the pyboard I could just put them on the flash drive and they got picked up from there. Is whatever those tools do to copy the files easily doable in bash?

**Reply by robert-hh** [reply]
mpremote is a command line tools. So you can easily call it from bash and put it into bash scripts.

**Comment by TheChymera** [comment]
@robert-hh ok, thanks for patiently explaining this :)
I guess I was fixated on just copying something in the shell based on my pyboard experience.
Anywho, I gave mpremote a try. Something's still wrong:

```
[dark]  ❱ ls RPI-RP2/
INDEX.HTM  INFO_UF2.TXT
[deco]  ❱ cp /tmp/rp2-pico-20230414-unstable-v1.19.1-1016-gb525f1c9e.uf2 RPI-RP2/
<drive auto-disconnects>
<unplug it>
<plug it back in *without* holding down boot>
<lights go on, but not the RBG light at the bottom, I assume this confirms something changed about it, but:> 
[deco]  ❱ mpremote
no device found
```

:(

Might be that my mpremote installation is broken, hard to check [without a test suite](https://github.com/orgs/micropython/discussions/11308).
Also, do I maybe require some kernel modules to be built for pyserial to properly work?

**Reply by robert-hh** [reply]
Do you see a /dev/ttyACMx device? Also, to get access to it, your user may have to be in the dialout group. That could be bypassed for a test by using sudo.

**Reply by TheChymera** [reply]
@robert-hh the device is there:

```
[deco]/tmp ❱ ls /dev/ttyACM*
/dev/ttyACM0
[deco]/tmp ❱ mpremote
no device found
```

But I don't have a dialout group on my system. Is that a group that means anything on linux? Never heard of it. Though it's clearly a permissions issue. `mpremote` run as root works. Should I just create the group? I don't have it present:

```
myhost ~ # gpasswd -a myuser dialup
gpasswd: group 'dialup' does not exist in /etc/group
```

**Reply by robert-hh** [reply]
The group name is `dialout`, not `dialup`. And you may have to use `sudo` to change the access rights.

**Reply by robert-hh** [reply]
What is the result of:
`ls -l /dev/ttyACM*`

**Reply by fdufnews** [reply]
> Is that a group that means anything on linux?

Yes of course. If you are part of the dialout group, you can use serial devices. Originally, it was giving access to modems hence the dialout name.

**Reply by TheChymera** [reply]
@robert-hh 

```
myhost ~ # ls -l /dev/ttyACM*
crw-rw---- 1 root dialout 166, 0 Apr 25 15:04 /dev/ttyACM0
```
But yes, adding my user to the dialout (not dialup :3 ) group fixed it. Thank you.

**Comment by TheChymera** [comment]
Thank you for your help. I think that almost wraps it up. Just one last issue now that I'm able to ~use it:

So I can transfer files to it with:

```
mpremote fs cp /path/on/computer.py :path/on/device.py
```

and edit them with 

```
mpremote edit /path/on/device.py
```

And re-start the device scripts with

```
mpremote reset
```

But how can I get stdout from the device? debugging is really difficult without stdout :(

**Reply by robert-hh** [reply]
Just calling:
mpremote
works a as terminal to get the REPL prompt from at the device. Then you can start a script with 
import scriptname.

You can as well run a script from your computer witth;

mpremote run scriptname

which will push the script to the device, start it, connects the stdin to your keyboard it and returns the output of the code.  Call `mpremote --help` for details.

**Reply by robert-hh** [reply]
The mount command is also very useful. e.g.

mpremote mount .

will  give you a REPL prompt with your computer's current directory mounted at the path /remote on the device.

**Reply by TheChymera** [reply]
Thank you! :)
